core/embed/rust/src/ui/flow/swipe.rs

```diff
@@ -1,29 +1,97 @@
 use crate::{
-    error,
+    error::{self, Error},
+    maybe_trace::MaybeTrace,
+    micropython::{
+        gc::{self, GcBox},
+        obj::Obj,
+    },
     ui::{
         component::{
             base::AttachType, swipe_detect::SwipeSettings, Component, Event, EventCtx, SwipeDetect,
             SwipeDetectMsg, SwipeDirection,
         },
+        display::Color,
         event::{SwipeEvent, TouchEvent},
-        flow::{base::Decision, FlowMsg, FlowState, FlowStore},
+        flow::{base::Decision, FlowMsg, FlowState},
         geometry::Rect,
-        shape::Renderer,
+        layout::obj::ObjComponent,
+        shape::{render_on_display, ConcreteRenderer, Renderer, ScopedRenderer},
         util::animation_disabled,
     },
 };
 
-/// Given a state enum and a corresponding FlowStore, create a Component that
-/// implements a swipe navigation between the states with animated transitions.
+use super::{base::StateChange, Swipable};
+
+use heapless::Vec;
+
+/// Component-like proto-object-safe trait.
+///
+/// This copies the Component interface, but it is parametrized by a concrete
+/// Renderer type, making it object-safe.
+pub trait FlowComponentTrait<'s, R: Renderer<'s>>: Swipable {
+    fn place(&mut self, bounds: Rect) -> Rect;
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<FlowMsg>;
+    fn render(&'s self, target: &mut R);
+
+    #[cfg(feature = "ui_debug")]
+    fn trace(&self, t: &mut dyn crate::trace::Tracer);
+}
+
+impl<'s, R, C> FlowComponentTrait<'s, R> for C
+where
+    C: Component<Msg = FlowMsg> + MaybeTrace + Swipable,
+    R: Renderer<'s>,
+{
+    fn place(&mut self, bounds: Rect) -> Rect {
+        <Self as Component>::place(self, bounds)
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<FlowMsg> {
+        <Self as Component>::event(self, ctx, event)
+    }
+
+    fn render(&'s self, target: &mut R) {
+        <Self as Component>::render(self, target)
+    }
+
+    #[cfg(feature = "ui_debug")]
+    fn trace(&self, t: &mut dyn crate::trace::Tracer) {
+        <Self as crate::trace::Trace>::trace(self, t)
+    }
+}
+
+/// Shortcut type for the concrete renderer passed into `render()` method.
+type RendererImpl<'a, 'alloc, 'env> = ScopedRenderer<'alloc, 'env, ConcreteRenderer<'a, 'alloc>>;
+
+/// Fully object-safe component-like trait for flow components.
+///
+/// This trait has no generic parameters:
+/// * it is instantiated with a concrete Renderer type, and
+/// * it requires the `FlowComponentTrait` trait to be implemented for _any_
+///   lifetimes.
```

Comment:
Trying to understand Higher Rank Trait Bounds ... here the `for` is used to erase lifetimes from a trait. The cost of this is that any function argument with such lifetime has the shortest possible lifetime, i.e. becomes invalid immediately after the function returns. Correct?

Reply:
That's a weird way to put it but essentially yes?
But note that this is the _default behavior_.
If you write a function accepting `impl Renderer`, without any lifetime parameters, the function also cannot rely on the argument existing beyond the scope of the function.

The issue here is that `FlowComponentTrait` has a lifetime parameter.
`Component` is fine with `render<'s>(&'s self, target: &mut impl Renderer<'s>)`, i.e., "render me with anything that lives as long as me"
But that's not object-safe. To be object-safe, we need to bind to a specific implementation of `Renderer` with a specific lifetime. That's the lifetime `'s` of `FlowComponentTrait`

buuuuut also, we `impl<'s> FlowComponentTrait<'s> for C: Component, R: Renderer<'s>`
so in other words, for any pre-specified `'s` and `R`, any `Component` implements `FlowComponentTrait` (which is obviously fine with Component, seeing as Component doesn't care about `'s`)

and the next step is `FlowComponentDynTrait`, saying "every implementer that implements `FlowComponentTrait<'s>` for any `'s` also implements `FlowComponentDynTrait`"
essentially circling back to what `Component` already does with `render<'s>(&'s self...)`